src/compose/comment_strip_iter.rs

```diff
@@ -1,90 +1,178 @@
-use std::{borrow::Cow, str::Lines};
+use std::{borrow::Cow, ops::Range};
+
+use winnow::{
+    ascii::till_line_ending,
+    combinator::{cut_err, opt},
+    error::StrContext,
+    token::any,
+    Located, PResult, Parser,
+};
+
+struct SourceCode {
+    /** Sorted pieces of the source code without any gaps */
+    parts: Vec<SourceCodePart>,
+}
+
+enum SourceCodePart {
+    Text(Range<usize>),
+    SingleLineComment(SingleLineComment),
+    MultiLineComment(MultiLineComment),
+}
+
+impl SourceCodePart {
+    fn span(&self) -> Range<usize> {
+        match self {
+            SourceCodePart::Text(span) => span.clone(),
+            SourceCodePart::SingleLineComment(comment) => comment.span.clone(),
+            SourceCodePart::MultiLineComment(comment) => comment.span.clone(),
+        }
+    }
+}
 
-use regex::Regex;
+pub struct SingleLineComment {
+    pub span: Range<usize>,
+}
+pub struct MultiLineComment {
+    pub span: Range<usize>,
+}
 
-static RE_COMMENT: once_cell::sync::Lazy<Regex> =
-    once_cell::sync::Lazy::new(|| Regex::new(r"(//|/\*|\*/)").unwrap());
+fn parse_source(input: &mut Located<&str>) -> PResult<SourceCode> {
+    let mut parts = Vec::new();
+    loop {
+        if input.is_empty() {
+            break;
+        }
+        if let Some(part) = opt(single_line_comment).parse_next(input)? {
+            parts.push(SourceCodePart::SingleLineComment(part));
+        } else if let Some(part) = opt(multi_line_comment).parse_next(input)? {
+            parts.push(SourceCodePart::MultiLineComment(part));
+        } else {
+            let text_span = any.span().parse_next(input)?;
+            if let Some(SourceCodePart::Text(last_span)) = parts.last_mut() {
+                last_span.end = text_span.end;
+            } else {
+                parts.push(SourceCodePart::Text(text_span));
+            }
+        }
+    }
+    Ok(SourceCode { parts })
+}
+
+fn single_line_comment(input: &mut Located<&str>) -> PResult<SingleLineComment> {
+    let start_span = "//".span().parse_next(input)?;
+    let text_span = till_line_ending.span().parse_next(input)?;
+    Ok(SingleLineComment {
+        span: start_span.start..text_span.end,
+    })
+}
+fn multi_line_comment(input: &mut Located<&str>) -> PResult<MultiLineComment> {
+    let start_span = "/*".span().parse_next(input)?;
+    loop {
+        if let Some(end_span) = opt("*/".span()).parse_next(input)? {
+            return Ok(MultiLineComment {
+                span: start_span.start..end_span.end,
+            });
+        } else if let Some(_) = opt(multi_line_comment).parse_next(input)? {
+            // We found a nested comment, skip it
+        } else {
+            // Skip a single character
+            let _ = cut_err(any)
+                .context(StrContext::Label("multiline comment"))
+                .parse_next(input)?;
+        }
+    }
+}
```

Comment:
Up until here is the actual parser code. Everything below is a dance to keep the existing behaviour mostly intact, while also prepping to replace more code.